File: Structures de donnees complexe/Tables de hacjages/Exercice 4/Implementation en Python/Food.py
from random import randrange
#Classe pour definir la structure de la date
class Date:
	#Fonction d'initialisation
	def __init__(self,jourval=None,moisval=None,anneeval=None):
		self.jour=jourval
		self.mois=moisval
		self.annee=anneeval
#Classe pour definir la structure d'un Repas 
class Repas:
	#Fonction d'initialisation
	def __init__(self,sexeval=None,ageval=None,date=None,heureval=None,repasval=None,uniteval=None,lieuval=None,nbpersval=None):
		self.sexe=sexeval
		self.age=ageval
		self.date=date
		self.heure=heureval
		self.repas=repasval
		self.unite=uniteval
		self.lieu=lieuval
		self.nbpers=nbpersval
#Classe pour definir la structure d'un noeud de la pile
class Noeud:
	#Fonction d'initialisation
	def __init__(self,elt=None):
		self.data=elt
		self.next=None
#Classe pour definir la structure de la pile
class File:
	#Fonction d'initialisation
	def __init__(self):
		self.tete=None
	#Fonction pour stocker un Food dans la pile
	def save_file(self,elt):
		node=Noeud()
		node.data=elt
		if (self.tete is None):
			self.tete=node
		else:
			tmp=self.tete
			while(tmp.next):
				tmp=tmp.next
			tmp.next=node
	#Fonction qui retourne le nombre d'element dans la pile
	def nbre(self):
		nb=1
		dernier=self.tete
		while(dernier.next):
			dernier=dernier.next
			nb+=1
		return nb
	#Fonction qui supprime un element de la pile a partir de son point de repas
	def supp_elt(self,repas):
		if self.tete is None:
			print("La self n'a aucun element")
		else :
			dernier=self.tete
			prec=self.tete
			i=0
			elt=dernier.data
			while (dernier.next)and(elt.repas!=repas):
				prec=dernier
				dernier=dernier.next
				elt=dernier.data
				i+=1
			if (elt.repas==repas)and(i>0):
				prec.next=dernier.next
			elif (elt.repas==repas)and(i==0):
				self.tete=dernier.next
	#Fonction qui retourne le Food de prix minimal dans la self
	def minimum(self):
		if self.tete is None:
			print("La self n'a aucun element")
			return None
		else:
			cour=self.tete
			mini=cour.data
			while(cour):
				elt=cour.data
				if(mini.prix>elt.prix):
					mini=elt
				cour=cour.next
			return mini
	#Fonction qui trie la self suivant les couts des Foods
	def sort_list(self):
		new=File()
		while(self.tete):
			mini=self.minimum()
			self.supp_elt(mini.repas)
			new.save_file(mini)
		return new
	#Fonction pour rechercher un element dans la pile
	def search_elt(self,repas):
		trouve=False
		tmp=self.tete
		elt=tmp.data
		while(tmp.next)and(elt.repas!=repas):
			tmp=tmp.next
			elt=tmp.data
		if (elt.repas==repas):
			trouve=True
		return trouve
	def print_list(self):
		if self.tete is None:
			print("self Vide...")
		else :
			tmp=self.tete
			i=1
			while (tmp is not None):
				elt=tmp.data
				print("{}\t{}\t{}\t{}\t{}".format(elt.heure,elt.repas,elt.lieu,elt.age,elt.nbpers))
				i+=1
				tmp=tmp.next
#Classe de la tabble de hachage
class TableHash :
	def __init__(self):
		self.tab=[]
		self.nbelt=4
		for i in range(self.nbelt):
			elt=File()
			self.tab+=[elt]
	def HashFunct(self,repas):
		return len(repas)%self.nbelt
	def save(self,elt):
		n=self.HashFunct(elt.repas)
		if self.tab[n] is None:
			liste=File()
			liste.save_file(elt)
			self.tab[n]=[liste]
		else :
			self.tab[n].save_file(elt)
	def read(self):
		x=["Masculin","Feminin"]
		y=["Maison","Ecole","Resto"]
		z=["Okok","Pate saute","Eru","Poisson","Poulet","Ndole","Nhui","Sauce tomate avec le riz"]
		w=["08h-12h","12h-16h","16h-18h"]
		#1
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#2
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#3
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#4
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#5
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#6
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#7
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#8
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#9
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
		#10
		date=Date(randrange(1,28),randrange(1,7),2022)
		elt=Repas(x[randrange(2)],randrange(15,25),date,w[randrange(3)],z[randrange(5)],"plat",y[randrange(3)],randrange(10))
		self.save(elt)
#Programme principal
table=TableHash()
table.read()
for i in range(table.nbelt):
	print("{} :".format(i+1))
	if table.tab[i].tete is not None :
		table.tab[i].print_list()
print("\n(C) 2022 By MMCJ/INF-L1")

File: Structures de donnees complexe/Tableaux/Exercice 4/Implementation en JS/Food.js
//Definition de la classe Date
class Date
{
	constructor (jour,mois,annee)
	{
		this.jour=jour;
		this.mois=mois;
		this.annee=annee;
	}
}
//Definition de la classe pour stocker les donnees de nourriture
class Repas{
	constructor (sexe,age,date,heure,repas,unite,lieu,nbpers)
	{
		this.sexe=sexe;
		this.age=age;
		this.date=date;
		this.heure=heure;
		this.repas=repas;
		this.unite=unite;
		this.lieu=lieu;
		this.nbpers=nbpers;
	}
}
class Table
{
	constructor ()
	{
		this.data= new Array();
		this.nbelt=0;
	}
	//Fonction pour stocker dans le tableau
	save(elt)
	{
		
		this.data[this.nbelt]=elt;
		this.nbelt=this.nbelt+1;
	}
	//Fonction pour lire les données
	read()
	{
		var dat,elt;
		dat =new Date(15,2,2022);
		elt=new Repas("Masculin",15,dat,"10h-14h","Okok","plat","Maison",6);
		this.save(elt);
		dat =new Date(13,3,2022);
		elt=new Repas("Feminin",10,dat,"14h-16h","Poulet","plat","Maison",9);
		this.save(elt);
		dat =new Date(15,5,2022);
		elt=new Repas("Masculin",18,dat,"16h-18h","Poisson","plat","Maison",1);
		this.save(elt);
		dat =new Date(15,7,2022);
		elt=new Repas("Masculin",25,dat,"08h-10h","Haricot","plat","Maison",3);
		this.save(elt);
		dat =new Date(10,8,2022);
		elt=new Repas("Masculin",13,dat,"18h-22h","Eru","plat","Resto",2);
		this.save(elt);
		dat =new Date(15,7,2022);
		elt=new Repas("Masculin",11,dat,"10h-14h","Koki","plat","Resto",3);
		this.save(elt);
		dat =new Date(15,7,2022);
		elt=new Repas("Masculin",28,dat,"08h-10h","Sanga","plat","Maison",4);
		this.save(elt);
		dat =new Date(15,7,2022);
		elt=new Repas("Masculin",23,dat,"08h-10h","Nkui","plat","Maison",7);
		this.save(elt);
	}
	//Permuter deux elements d'une liste
	echange(i,j)
	{
		var c;
		c=this.data[i];
		this.data[i]=this.data[j];
		this.data[j]=c;
	}
	sort_list()
	{
		var i,j;
		for (i=0;i<this.nbelt-1;i++)
		{
			var mini=i;
			for (j=i+1;j<this.nbelt;j++)
			{
				if (this.data[j].age<this.data[mini].age)
				{
					mini=j;
				}
			}
			this.echange(i,mini);
		}
	}
	seqSearch(elt)
	{
		var trouve=false;
		i=0;
		while(this.data[i].repas!=elt)and(i<this.nbelt)
		{
			if (this.data[i].repas===elt)
			{
				trouve=true;
			}
			i++;
		}
		return trouve;
	}
	print_list()
	{
		var i,elt;
		for(i=0;i<this.nbelt;i++)
		{
			elt=this.data[i];
			console.log(elt.repas+"\t"+elt.lieu+"\t"+elt.age+"\t"+elt.nbpers);
		}
	}
}
var tab= new Table();
tab.read();
console.log("\nTaille \n"+tab.nbelt);
tab.print_list();
tab.sort_list();
console.log("\nTaille \n"+tab.nbelt);
tab.print_list();
